main.go

```diff
@@ -15,29 +15,89 @@ var m = make(map[int]contact)
 
 func main() {
 	r := mux.NewRouter()
-	r.HandleFunc("/api/v1/contants", getContantsHandler).Methods("GET")
+	//r.HandleFunc("/api/v1/contants", getContantsHandler).Methods("GET")
 	r.HandleFunc("/api/v1/contants", postContantsHandler).Methods("POST")
 	r.HandleFunc("/api/v1/contants/{id}", deleteContantsHandler).Methods("DELETE")
+	r.HandleFunc("/api/v1/contants/{id}", getByIdContantsHandler).Methods("GET")
+	r.HandleFunc("/api/v1/contants/{id}", putContantsHandler).Methods("PUT")
 	http.ListenAndServe(":8080", r)
 }
 
-func getContantsHandler(w http.ResponseWriter, r *http.Request) {
-	contList := make([]contact, 0)
-	searchKey := r.URL.Query().Get("search")
-	for _, cont := range m {
-		if searchKey != "" {
-			if strings.Contains(cont.Name, searchKey) {
-				contList = append(contList, cont)
-			}
-		} else {
-			contList = append(contList, cont)
-		}
+func getByIdContantsHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	params := mux.Vars(r)
+	id := params["id"]
+	contantID, err := strconv.Atoi(id)
+	if err != nil {
+		errMsg := fmt.Sprintf("Could not convert value of id to integer: %v. Are passing non int value?", id)
+		log.Println(errMsg, err.Error())
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+	_, ok := m[contantID]
+	if !ok {
+		errMsg := fmt.Sprintf("Could not found contant: %v. Is it not present?", id)
+		log.Println(errMsg)
+		http.Error(w, errMsg, http.StatusGone)
+		return
 	}
-	data, _ := json.Marshal(contList)
+	
+	var cont contact = m[contantID]
+	 data, _ := json.Marshal(cont)
 	w.Header().Set("Content-Type", "application/json")
 	w.Write(data)
 }
 
+func putContantsHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	params := mux.Vars(r)
+	id := params["id"]
+	contantID, err := strconv.Atoi(id)
+	if err != nil {
+		errMsg := fmt.Sprintf("Could not convert value of id to integer: %v. Are passing non int value?", id)
+		log.Println(errMsg, err.Error())
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+
+	_, ok := m[contantID]
+	if !ok {
+		errMsg := fmt.Sprintf("Could not found contant: %v. Is it not present?", id)
+		log.Println(errMsg)
+		http.Error(w, errMsg, http.StatusGone)
+		return
+	}
+	var cont contact
+	errr := json.NewDecoder(r.Body).Decode(&cont)
```

Comment:
you should decode this before getting it from the map. at line: https://github.com/Prithvipal/phone-dir/pull/8/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R62